.changeset/empty-fans-reflect.md

```diff
@@ -0,0 +1,65 @@
+---
+'astro': minor
+---
+
+Adds a new `experimental.typescript` configuration option
+
+If you enable this option, Astro will generate a `tsconfig.json` file under `./.astro/`:
+
+```diff
+import { defineConfig } from "astro/config"
+
+export default defineConfig({
++    experimental: {
++        typescript: {}
++    }
+})
+```
+
+It will create references under the hood, making `env.d.ts` optional if you haven't made any change to it:
+
+```diff
+-/// <reference types="astro/client" />
+-/// <reference path="../.astro/types.d.ts" />
+```
+
+Enabling this option will require a few updates to your root `tsconfig.json`:
+
+1. You'll need to make sure `extends` points to `./.astro/tsconfig.json`. If it's not correctly set, an eror will be thrown with a diff to help you
+2. If you have `include` or `exclude`, you'll have to move them to respectively `experimental.typescript.include` and `experimental.typescript.exclude`:
+
+```diff
+{
+-    "include": ["foo"],
+-    "exclude": ["bar"]
+}
+```
+
+```diff
+import { defineConfig } from "astro/config"
+
+export default defineConfig({
+    experimental: {
+        typescript: {
++            include: ["foo"],
++            exclude: ["bar"]
+        }
+    }
+})
+```
```

Comment:
There is a small typo on `error` (`eror` > `error`) that needs to be fixed. Other than that, I'm not Sarah 😁 but I think this should a little reworded to be more direct and more precise (the two diff blocks):
````suggestion
Enabling this option will require a few updates to your root `tsconfig.json`:

1. `extends` must point to `./.astro/tsconfig.json` otherwise an error will be thrown with a diff to help you.
2. If set, move `include` and `exclude` to `experimental.typescript.include` and `experimental.typescript.exclude` respectively in your Astro config file.

For example, in `tsconfig.json` you'll need to make the following changes:

```diff
{
+    "extends": "./.astro/tsconfig.json"
-    "include": ["foo"],
-    "exclude": ["bar"]
}
```

Then you'll need to reflect them in `astro.config.mjs`:

```diff
import { defineConfig } from "astro/config"

export default defineConfig({
    experimental: {
        typescript: {
+            include: ["foo"],
+            exclude: ["bar"]
        }
    }
})
```
````

I'm not sure about the ``in `astro.config.mjs`:`` since we can have different extensions (e.g. `.ts`) but this will help the user to understand where these diff take place.

But again, I'm not Sarah or Chris... so maybe you'll have better suggestions. 😄